Get aggregated cost report

vihuang93 opened the pull request:
Postman collection: https://www.getpostman.com/collections/918c7e6b50da8c5482a2

Three endpoints added:
GET ("/costs/{id}") by show Id
This will return an aggregated cost report. This endpoint should work for both non-episodic or episodic shows.

POST ("/costs")
This will take in a json object of ShowEpisodeCostAPIModel, and persist a row in SHOW_EPISODE_COST table.

GET ("/prodcosts/{id}") by show Id
This will return an aggregated cost report. This endpoint will add proportioned amortized costs. It should work for both non-episodic or episodic shows.

Created db:
<img width="269" alt="Screen Shot 2021-05-26 at 10 59 14 PM" src="https://user-images.githubusercontent.com/74227094/119764144-80210c80-be76-11eb-9859-e32e8aeca9a6.png">
Tested in postman: Get aggregated cost report for show 1
<img width="733" alt="Screen Shot 2021-05-26 at 10 59 28 PM" src="https://user-images.githubusercontent.com/74227094/119764161-87e0b100-be76-11eb-9891-d6df817b6254.png">
Tested in postman: Get aggregated cost report for show 2
<img width="705" alt="Screen Shot 2021-05-26 at 10 59 48 PM" src="https://user-images.githubusercontent.com/74227094/119764175-8dd69200-be76-11eb-98d6-f8e4bba27894.png">

GET production costs including amortized costs per season:
DB: 
<img width="878" alt="Screen Shot 2021-05-27 at 12 24 53 AM" src="https://user-images.githubusercontent.com/74227094/119770682-258dad80-be82-11eb-92cc-117cbaae7522.png">
GET production costs for show id = 2
-- for movie 2, amortized cost per episode is 100,
-- Episode 101: 400 + 100 = 500
-- Episode 102: 300 + 100 = 400
-- Episode 103: 200 + 100 = 300
<img width="724" alt="Screen Shot 2021-05-27 at 12 25 09 AM" src="https://user-images.githubusercontent.com/74227094/119770725-34746000-be82-11eb-9520-7f5b9de6ad09.png">

**corner cases**
1. POST creation, if no record created, it should return 200 instead of 201.
<img width="932" alt="Screen Shot 2021-05-27 at 10 00 46 PM" src="https://user-images.githubusercontent.com/74227094/119923267-02710580-bf37-11eb-8e2a-add1e5aaa800.png">

2. GET operation, if no record found, it should retun 404.
<img width="845" alt="Screen Shot 2021-05-27 at 8 58 28 PM" src="https://user-images.githubusercontent.com/74227094/119923220-eec59f00-bf36-11eb-87be-6777b756224b.png">

**“bad data”**
POST creation, unrecognized request body calling POST, return 400 bad request 'episode_co' is not a valid field
<img width="847" alt="Screen Shot 2021-05-27 at 10 02 46 PM" src="https://user-images.githubusercontent.com/74227094/119923443-51b73600-bf37-11eb-9c0f-d1bfbbe977f6.png">


POST creation, Episode code length != 3, , return 400 bad request [{ "id":"4", "episode_code":"1111", "amount":"200" } ]
<img width="868" alt="Screen Shot 2021-05-27 at 10 03 24 PM" src="https://user-images.githubusercontent.com/74227094/119923489-65629c80-bf37-11eb-9fe8-c93b780e9a5a.png">

POST creation, request object missing field [{

"episode_code":"1111", "amount":"200" } ]
<img width="895" alt="Screen Shot 2021-05-27 at 10 03 55 PM" src="https://user-images.githubusercontent.com/74227094/119923527-74e1e580-bf37-11eb-906e-b41b64381082.png">

Commit:
Cost Reports - initial commits - add DAO, data model, repository and api structures.

Commit:
Cost Reports - initial commits - remove data access module. Add api models

Commit:
Cost Reports - fix dao issue. add computation logic to generate aggregated cost report

Commit:
Cost Reports - add api for amortized cost

Commit:
Cost Reports - update java doc

Commit:
Cost Reports - POST a list instead of a single object

Commit:
Cost Reports - add unit tests

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/CostmanagementservApplication.java:
```diff
@@ -0,0 +1,130 @@
+
+@SpringBootApplication
+@RestController
+public class CostmanagementservApplication {
```
Should we look at adding some logging to this application?

vihuang93 replied:
We definitely can. We can add logging on requests as well as exceptions. Didn't think of adding any thinking this is a small practice.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/CostmanagementservApplication.java:
```diff
@@ -0,0 +1,130 @@
+		ObjectMapper objectMapper = new ObjectMapper();
+		List<ShowEpisodeCostAPIModel> showEpisodeCostList;
+		try {
+			showEpisodeCostList = objectMapper.readValue(showEpisodeCostListStr, new TypeReference<List<ShowEpisodeCostAPIModel>>() {
```
Any specific reason to do the explicit conversion from json string to pojo or can we let Spring boot handle this?

vihuang93 replied:
good question.. originally tried passing ShowEpisodeCostAPIModel and use spring annotations for validation, but during testing, I realize the annotations are not working, therefore forced it to do the explicit conversion here.

vihuang93 replied:
I reverted back to just take the api model

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/CostmanagementservApplication.java:
```diff
@@ -0,0 +1,130 @@
+		List<ShowEpisodeCostAPIModel> createdList = new ArrayList<>();
+		for(ShowEpisodeCostAPIModel showEpisodeCostAPIModel:showEpisodeCostList) {
+			if(showEpisodeCostAPIModel.getEpisode_code() == null || showEpisodeCostAPIModel.getAmount() == null
+					|| showEpisodeCostAPIModel.getId() == null || showEpisodeCostAPIModel.getEpisode_code().length() != 3){
```
Apart from episode length, what other validations can we use to guard against a bad episode codes?
Some bad examples that will bypass the length validation -
1. Season 2 of a show has 5 episodes but the episode code in the input is 208
2. Season 2 of a show having an episode code of 402.

vihuang93 replied:
With our current design, there won't be a way to check for those scenarios. we don't know if an episode does exists.
  SHOW_EPISODE_COST only keeps each cost transaction for an episode. 

To satisfy the error scenarios you mentioned, I would suggest to have some extra tables and apis.
SHOW_SEASON TABLE:
ID(PK), SHOW_ID, SEASON_NUMBER, description(if any, or other metadata)
SHOW_EPISODES:
ID(PK), season_ID(ID of show_season), episode_number, season_number, description(if any, or other metadata)
whenever a new episode/season is created, appropriate row should be inserted. that should be done with other apis.
Based on above,
Scenario 1. Season 2 of a show has 5 episodes but the episode code in the input is 208
request will still be {
	"show_id":
	"episode_code":
	"cost":
}
before we insert in show_episode_cost, we can have a validation to check if this episode_code is valid. select * from SHOW_EPISODES where episode_number = ? and season_number = ?. if this returned zero, meaning this episode_code is invalid. then error out.

Scenario 2, we won't be able to check for this scenario. actually I am unclear on this scenario.  when we are given 402, meaning client wants to create a cost for season 4 episode 2, we will only check if this episode exists. If exists, persist, if not, error out. In this case, if season4 episode 2 exists, I will add this cost for 402. It should be the client's responsibility to give valid cost data. (i imagine there will be UI to select valid season and episode before adding cost, and episode_code should be generated then passed to our service)

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/CostmanagementservApplication.java:
```diff
@@ -0,0 +1,130 @@
+					|| showEpisodeCostAPIModel.getId() == null || showEpisodeCostAPIModel.getEpisode_code().length() != 3){
+				return ResponseEntity.badRequest().build();
+			}
+			Optional<ShowEpisodeCost> createdEpisodeCost = this.costsDbRepository.createCost(Long.parseLong(showEpisodeCostAPIModel.getId()), showEpisodeCostAPIModel.getEpisode_code(),
```
This might be slow if there are many records for each show and also many shows for which we are calling this API in one shot for record insertion. Any way to make it more performant?

vihuang93 replied:
we can use batch insertion on the db side to reduce db connection costs. depend on how many records we will need to insert in each request, if say we have to insert 1000 in a request, we can break it down into smaller chunks before doing batch insertion. this number can vary. let's say we do 100 in one batch insertion, then we call batchInsert 10 times for this 1000 records. reason to break it down is because 100 per insertion is transactional, meaning one fail, all fail. this is to reduce chances of failing. also you might have db restrictions on how many rows you can commit each time.

vihuang93 replied:
I updated the create to use batch insert. Assumption here is the list I get is < 100. If we have thousands, we will need to break it down to smaller chunks.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/CostmanagementservApplication.java:
```diff
@@ -0,0 +1,130 @@
+				createdEpisodeCostApi.setAmount(String.valueOf(createdEpisodeCost.get().getAmount()));
+
+				createdList.add(createdEpisodeCostApi);
+			}
```
What happens if we have to insert 1000 records and If for some reason the object creation failed for 1 record say due to some transient database error? What is a graceful way to handle that situation?

vihuang93 replied:
in one of the previous comment, I suggested doing batch insertion. 1000 records can be broken into 10 batch inserts. if one of batch insertion failed(batch insert 1-100, but record 10 failed), we could do retry, let's say we do 3. If retry failed, we can return back the records that had issue. in this example case, 900 succeed, 100 failed, we tell client this 100 are not persisted. in error details we can give more info on which specific record had issue so the client can look into that.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/apimodels/ShowEpisodeCostAPIModel.java:
```diff
@@ -0,0 +1,45 @@
+
+public class ShowEpisodeCostAPIModel {
+    @NotNull
+    String id;
```
Is show id or movie id a more descriptive name for this field?

vihuang93 replied:
yeah, initially i was going to create id(auto generated id for each row) & show id. Will say we will keep this one and add a new Column show_id.

vihuang93 replied:
updated this id to showId, added another auto generated column ID as the primary key.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/apimodels/ShowEpisodeCostAPIModel.java:
```diff
@@ -0,0 +1,45 @@
+    String episode_code;
+
+    @NotNull
+    String amount;
```
Seeing some conversion between string and Long for amount throughout the code. Can this just be long type instead?

vihuang93 replied:
updated.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/datamodels/ShowEpisodeCost.java:
```diff
@@ -0,0 +1,55 @@
+ */
+@Entity
+@Table(name = "SHOW_EPISODE_COSTS")
+public class ShowEpisodeCost {
```
Would it be useful to have an additional primary key column for each show episode cost object that uniquely identifies each record?
What kind of additional operations would a primary key help support?

vihuang93 replied:
yeah, we should always have an id for each row in real life. the primary key is not useful here for our requirement but it will be useful if we want to do update on specific rows.

vihuang93 replied:
added

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/repository/CostsDbRepository.java:
```diff
@@ -0,0 +1,121 @@
+     * proportion of amortized cost = total amortized costs for a season/episodes of a season
+     */
+    public List<ShowEpisodeCost> getProductionCostsIncludingAmortizedCost(Long id){
+        List<ShowEpisodeCost> list = showEpisodeCostDAO.findAllEpisodeCostsByShowId(id);
```
List as the variable name seems like a little confusing here. Something like allEpisodeCostsForShow would be a better name?

vihuang93 replied:
yup, updated.

ananyarnflx commented on costmanagementserv/src/main/java/com/costs/costmanagement/repository/CostsDbRepository.java:
```diff
@@ -0,0 +1,121 @@
+        for(Map.Entry<Integer, Long> entry:seasonAmortizedCostMap.entrySet()){
+            int season = entry.getKey();
+            long totalAmortizedCost = entry.getValue();
+            long amortizedCostPerEpisode = totalAmortizedCost/seasonEpisodeCountMap.get(season);
```
What if the amortized cost is such that it can't be evenly divided by the number of episodes? i.e. total cost of 98 to be divided among 9 episodes?

vihuang93 replied:
I can't make a decision on that, this is a corner case we will need to clarify from the product manager, how many digit after decimal number should we keep？ if cannot be divided, should we 1) round up, or 2) is it ok for one episode to own more. for example totalAmortizedCost = 10, total episode 3, if we were to keep 1 digits after decimal, is it ok to do (3.4, 3.3 and 3.3). in which case I would be more towards to second option

vihuang93 commented:
@ananyarnflx Ack. will look into the comments.

Commit:
costs report - address comments

vihuang93 commented:
@ananyarnflx Thanks for taking the time reviewing. Sorry about the late update here. I am on call this week so didn't have much time to work on the pr earlier. 
I addressed most of the comments and pushed the changes. There are places I replied to with my thoughts but didn't make actual changes due to time limitations, hopefully the reply makes sense. If not, please let me know and we can have further discussions.

ananyarnflx commented:
Thanks Vinnie. Hope you enjoyed the exercise!
